src/replit_river/codegen/client.py

```diff
@@ -798,10 +806,314 @@ def __init__(self, client: river.Client[{handshake_type}]):
+                )
+            ]
+        )
+    elif procedure.type == "upload":
+        assert output_meta
+        assert error_meta
+        _, output_type, parse_output_method = output_meta
+        _, error_type, parse_error_method = error_meta
+        error_type_name = extract_inner_type(error_type)
+
+        output_or_error_type = UnionTypeExpr([output_type, error_type_name])
+
+        if init_meta and input_meta:
+            _, init_type, render_init_method = init_meta
+            _, input_type, render_input_method = input_meta
+            current_chunks.extend(
+                [
+                    reindent(
+                        "  ",
+                        f"""\
+        async def {name}(
+          self,
+          init: {render_type_expr(init_type)},
+          inputStream: AsyncIterable[{render_type_expr(input_type)}],
+        ) -> {render_type_expr(output_type)}:
+          return await self.client.send_upload(
+            {repr(schema_name)},
+            {repr(name)},
+            init,
+            inputStream,
+            {reindent("                    ", render_init_method)},
+            {reindent("                    ", render_input_method)},
+            {reindent("                    ", parse_output_method)},
+            {reindent("                    ", parse_error_method)},
+          )
+                    """,
+                    )
+                ]
+            )
+        elif protocol_version == "v1.1":
+            assert input_meta, "Protocol v1 requires input to be defined"
+            _, input_type, render_input_method = input_meta
+            current_chunks.extend(
+                [
+                    reindent(
+                        "  ",
+                        f"""\
+        async def {name}(
+          self,
+          inputStream: AsyncIterable[{render_type_expr(input_type)}],
+        ) -> {  # TODO(dstewart) This should just be output_type
+                            render_type_expr(output_or_error_type)
+                        }:
+          return await self.client.send_upload(
+            {repr(schema_name)},
+            {repr(name)},
+            None,
+            inputStream,
+            None,
+            {reindent("                    ", render_input_method)},
+            {reindent("                    ", parse_output_method)},
+            {reindent("                    ", parse_error_method)},
+          )
+                    """,
+                    )
+                ]
+            )
+        elif protocol_version == "v2.0":
+            assert init_meta, "Protocol v2 requires init to be defined"
+            _, init_type, render_init_method = init_meta
+            current_chunks.extend(
+                [
+                    reindent(
+                        "  ",
+                        f"""\
+        async def {name}(
+          self,
+          init: {render_type_expr(init_type)},
+        ) -> {  # TODO(dstewart) This should just be output_type
```

Comment:
hm why do we want output_type over output_or_error_type

Reply:
In the current world where we throw exceptions all over the place, this muddies the water. I'd like to consolidate, but post v2.